### PR TITLE
Fix e2e testing and (hopefully) all routing issues

### PR DESCRIPTION
**Pull request by Var1377**
Hopefully this works, it also fixes login redirect and sign-out redirects

### CODE REVIEW

**Comment by Var1377** [comment]
Fixes auth flow but breaks all the tests...

File: cypress/e2e/auth.cy.ts
Comment (DanWilkins2107) [inline cypress/e2e/auth.cy.ts:9]: Should this now be removed?

**Review by DanWilkins2107 — APPROVED** [review]

File: src/components/NavBar/SignOutButton.tsx
Comment (SamYap-Softwire) [inline src/components/NavBar/SignOutButton.tsx:1]: Why do we need "use client" here?

File: src/app/layout.tsx
Comment (SamYap-Softwire) [inline src/app/layout.tsx:9]: Can you explain why this is `React.ReactNode` instead of `React.ReactElement`?

**Review by SamYap-Softwire — COMMENTED** [review]
Seems good. Will approve once my questions are answered :)

File: src/components/NavBar/SignOutButton.tsx
Comment (Var1377) [reply src/components/NavBar/SignOutButton.tsx:1]: I think we need it for `createClientComponentClient`, I'm not totally sure though I'll have a look into it

File: src/app/layout.tsx
Comment (Var1377) [reply src/app/layout.tsx:9]: React.ReactNode is the type we should be using for children. React.ReactElement means we can't use strings/numbers etc. and is causing unneccessary fragments across the codebase

File: cypress/e2e/accessibility.cy.ts
Comment (SamYap-Softwire) [inline cypress/e2e/accessibility.cy.ts:35]: What does `data-loaded="true"` check for?

File: cypress/e2e/auth.cy.ts
Comment (SamYap-Softwire) [inline cypress/e2e/auth.cy.ts:44]: Is it worth checking that `cy.visit("/clients")` actually redirects to clients page before clicking the sign out button? 
(in case it stays on the login page despite `cy.visit("/clients")`)

File: cypress/e2e/auth.cy.ts
Comment (SamYap-Softwire) [inline cypress/e2e/auth.cy.ts:53]: Is this in a sign out state initially?

**Review by daniel-conlan-softwire — CHANGES_REQUESTED** [review]
Looks good in general, just a small bug in the dark mode accessibility tests I believe

File: cypress/e2e/accessibility.cy.ts
Comment (SamYap-Softwire) [inline cypress/e2e/accessibility.cy.ts:78]: Do we need `cy.get("label[aria-label='Theme Switch']").click();` here?

File: cypress/support/commands.ts
Comment (SamYap-Softwire) [inline cypress/support/commands.ts:25]: Can you explain what the `$` is for?

File: cypress/support/commands.ts
Comment (SamYap-Softwire) [inline cypress/support/commands.ts:18]: Why 5?

File: cypress.config.ts
Comment (Var1377) [reply cypress.config.ts:34]: So we know why tests fail. You can get to them by clicking on checks and scrolling down to artifacts.

File: src/components/DataInput/DataInput.cy.tsx
Comment (SamYap-Softwire) [inline src/components/DataInput/DataInput.cy.tsx:170]: Why do we need this?

File: src/components/LoginPanel.tsx
Comment (SamYap-Softwire) [inline src/components/LoginPanel.tsx:43]: Why do we need this? Is it different from just setting `data-loaded={true}`?

**Review by SamYap-Softwire — CHANGES_REQUESTED** [review]
Sorry, quite a few things that I don't understand and need clarification for before I approve this. 👉 👈

File: cypress/e2e/auth.cy.ts
Comment (Var1377) [reply cypress/e2e/auth.cy.ts:5]: The issues between locally testing and testing on actions are partially down to the performance difference, especially in terms of compiling the page on very first load and in terms of time until hydration. This just accounts for this.

File: cypress/e2e/accessibility.cy.ts
Comment (Var1377) [reply cypress/e2e/accessibility.cy.ts:35]: `useEffect` only runs after hydration, at which point the data-loaded attribute is set to true. With a long timeout, this has the effect of waiting until hydration happens, and means we don't have to wait an arbitrary amount of time and hoping that it'll have hydrated by then.

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:25]: This is a jquery idiom (cypress uses it under the hood). I'm not sure if we want to keep them or not, but this is generally what jquery code looks like.

https://jquery.com/
(jquery was a very commonly used library that is now largely redundant due to the improvement of browsers)

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:18]: It was arbitrary, I just wanted it to happen within 10 seconds although 0.5 + 1 + 2 + 4 + 8 > 10 so I did my maths wrong ;-;

File: src/components/DataInput/DataInput.cy.tsx
Comment (Var1377) [reply src/components/DataInput/DataInput.cy.tsx:170]: This test breaks on actions about 1/8 times, adding a redundant get statement adds some more time and hopefully means reliability improves.

File: src/components/LoginPanel.tsx
Comment (Var1377) [reply src/components/LoginPanel.tsx:43]: data-loaded={true} would mean on a server-side-rendered document, it would still be true. The key thing here is that `useEffect` only runs after hydration, so `loaded === true` implies that hydration has happened and themeing + routing infrastructure should be in place.

File: src/components/DataInput/DataInput.cy.tsx
Comment (SamYap-Softwire) [reply src/components/DataInput/DataInput.cy.tsx:170]: Is there no other way to do this?

File: cypress/support/commands.ts
Comment (SamYap-Softwire) [reply cypress/support/commands.ts:25]: I think I prefer the variable naming to be clearer and without the `$` sign.

File: cypress/support/commands.ts
Comment (DanWilkins2107) [inline cypress/support/commands.ts:91]: How come we're not using `cy.checkAccessibility`

File: src/components/DataInput/DataInput.cy.tsx
Comment (DanWilkins2107) [inline src/components/DataInput/DataInput.cy.tsx:175]: So my understanding of this is wrong I think, but I thought this div targeted like the header bit of the dropdown menu. If this is the case, would this not open and shut the menu? Does seem to work though so I think I'm wrong.

**Review by DanWilkins2107 — APPROVED** [review]
A few questions, don't see any major issues though

File: cypress/e2e/accessibility.cy.ts
Comment (Var1377) [reply cypress/e2e/accessibility.cy.ts:3]: Because we don't want it to run when checking the login page and it's better than logging in and the signing out as that takes more time

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:91]: It's used in accessibility.cy.ts

File: src/components/DataInput/DataInput.cy.tsx
Comment (Var1377) [reply src/components/DataInput/DataInput.cy.tsx:170]: There probably is, but I'm personally happy with this solution as it's concern is that the `Change handler for dropdown works`. The current issue is an asynchronously fired DOM change and not an issue with the dropdown. Feel free to experiment around with another solution, although it's tough to know what works when the failure happens so infrequently

Comment (Var1377) [reply src/components/DataInput/DataInput.cy.tsx:170]: I also don't know if it's actually a solution, it's tough to tell but I haven't had any test failures on this specific test since I added this.

Comment (Var1377) [reply src/components/DataInput/DataInput.cy.tsx:170]: Just experimented with the dropdown, and when you click a list item it closes. Not sure how the other test was passing but this is how the interaction would happen

File: src/components/NavBar/SignOutButton.tsx
Comment (Var1377) [reply src/components/NavBar/SignOutButton.tsx:1]: Yep we do :)

**Review by daniel-conlan-softwire — APPROVED** [review]

File: cypress.config.ts
Comment (TomCusack-Softwire) [inline cypress.config.ts:34]: did you figure out how to set a really small retention period? (e.g. delete after a day)

File: cypress/e2e/auth.cy.ts
Comment (TomCusack-Softwire) [inline cypress/e2e/auth.cy.ts:2]: is `: string` really needed? can this just be automatically inferred (potentially with a `!` in case of null)?

File: cypress/support/commands.ts
Comment (TomCusack-Softwire) [inline cypress/support/commands.ts:21]: I see this being very easily missed - is there a way to make it a warning?

File: cypress/support/commands.ts
Comment (TomCusack-Softwire) [inline cypress/support/commands.ts:29]: is this the same as `submitButton.click()`?

File: cypress/e2e/auth.cy.ts
Comment (TomCusack-Softwire) [inline cypress/e2e/auth.cy.ts:49]: is this not the same as `Redirected to login page` tests? especially if we don't start off in a logged in state

File: cypress/e2e/auth.cy.ts
Comment (TomCusack-Softwire) [inline cypress/e2e/auth.cy.ts:25]: why sometimes `#email` and sometimes `input#email`?

Comment (TomCusack-Softwire) [reply cypress/e2e/auth.cy.ts:25]: can this be 'better' semantically? e.g. testing for the thing that has 'email' in the label - idk what you can access

File: src/app/auth.tsx
Comment (TomCusack-Softwire) [inline src/app/auth.tsx:21]: can these be combined to `if (loggedIn && ...) return; if (!loggedIn && ...) return;`? I hate else

File: src/app/auth.tsx
Comment (TomCusack-Softwire) [inline src/app/auth.tsx:1]: are there any issues with doing auth 'client side' - can a malicious client control this? assuming not?

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]

File: cypress.config.ts
Comment (Var1377) [reply cypress.config.ts:34]: yep, it's what the `retention-days` key does in `checks.yml`

File: cypress/e2e/auth.cy.ts
Comment (Var1377) [reply cypress/e2e/auth.cy.ts:2]: It's inferred as any without it (which is fine, it's just nice to have everything be typed)

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:29]: No, if the button is disabled and we try `submitButton.click()`, it would result in a test failure.

File: cypress/e2e/auth.cy.ts
Comment (Var1377) [reply cypress/e2e/auth.cy.ts:49]: No, because we wait for hydration and click `Link`s.

This ensures it uses client side routing and covers our bases better. With how it's currently implemented there shouldn't be a difference, but having the test ensures it stays that way.

File: src/app/auth.tsx
Comment (Var1377) [reply src/app/auth.tsx:1]: Middleware and database level security still protect us from those without authentication. The one thing that concerns me about a malicious client is whether sensitive data remains in cache after an authenticated user has logged out although I'm not sure if that's something we should be concerned with.

File: src/app/auth.tsx
Comment (TomCusack-Softwire) [inline src/app/auth.tsx:22]: I'd prefer 
```ts
if (loggedIn && ...("/login")) {
    return "/clients";
}
if (!loggedIn && ... "/login") {
    return "/login";
}
return null;
```
just reads a _lot_ better in my opinion

Comment (TomCusack-Softwire) [reply src/app/auth.tsx:22]: very similar to how I wrote: https://github.com/VauxhallFoodbank/vauxhall-foodbank/blob/a6347c23f729be7903dd0192438f3214795a0dfa/src/middleware.ts#L13

**Review by TomCusack-Softwire — APPROVED** [review]

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:29]: submitButton is a jquery element, and `click` is deprecated in jquery in favour of `.trigger("click")`

I could also do `cy.wrap(submitButton).click()` if you prefer that?

File: cypress/support/commands.ts
Comment (Var1377) [reply cypress/support/commands.ts:21]: Unfortunarely cy.warn isn't a thing, it's either this or console.warn but the latter isn't guaranteed to be displayed.

**Pull request merged**
